src/main/java/com/linkedin/kmf/services/ProduceService.java

```diff
@@ -154,15 +156,44 @@ public void start() {
   }
 
   private void initializeStateForPartitions() {
+    Map<Integer, String> keyMapping = generateKeyMappings();
     int partitionNum = _partitionNum.get();
     for (int partition = 0; partition < partitionNum; partition++) {
+      String key = keyMapping.get(partition);
       //This is what preserves sequence numbers across restarts
-      if (!_nextIndexPerPartition.containsKey(partition)) {
-        _nextIndexPerPartition.put(partition, new AtomicLong(0));
+      if (!_nextIndexPerKey.containsKey(key)) {
+        _nextIndexPerKey.put(key, new AtomicLong(0));
         _sensors.addPartitionSensors(partition);
       }
-      _produceExecutor.scheduleWithFixedDelay(new ProduceRunnable(partition), _produceDelayMs, _produceDelayMs, TimeUnit.MILLISECONDS);
+      _produceExecutor.scheduleWithFixedDelay(new ProduceRunnable(partition, key), _produceDelayMs, _produceDelayMs, TimeUnit.MILLISECONDS);
+    }
+  }
+
+  private Map<Integer, String> generateKeyMappings() {
+    int partitionNum = _partitionNum.get();
+    HashMap<Integer, String> keyMapping = new HashMap<>();
+
+    int keyIndex = 0;
+    while (keyMapping.size() < partitionNum) {
+      String key = keyIndex + "";
+      int partition = getDefaultPartitionForKey(key);
+      if (!keyMapping.containsKey(partition)) {
+        keyMapping.put(partition, key);
+      }
+      keyIndex++;
     }
+
+    return keyMapping;
+  }
+
+  //TODO: Remove method and convert uses to Util.toPositive when kafka version dependency reaches 0.10.1.10
+  public static int toPositive(int number) {
+    return number & 0x7fffffff;
+  }
+
+  private int getDefaultPartitionForKey(String key) {
```

Comment:
Yeah, I was thinking the same thing but found that the current design of partitioners would require us to base pipeline monitoring design decisions on partitioner implementations details.

Options would be to:
1) Base our design on partitioner implementations details
2) Create key hashing classes in monitor that correspond to hashing strategies in kafka
3) Create key hashing classes in kafka and have partitioners import them and pipeline monitor import them

Option 3 has is the cleanest solution but has the broadest implications. Option 2 may be a good compromise between the options.